src/query/shorthand.ts

```diff
@@ -277,3 +282,205 @@ export function fieldDef(encQ: EncodingQuery,
+    }
+
+    specQ.encodings.push(encQ);
+  }
+
+  export function rawFieldDef(encQ: EncodingQuery, fieldDefPart: string[]): EncodingQuery {
+    encQ.field = fieldDefPart[0];
+    encQ.type = TYPE_FROM_SHORT_TYPE[fieldDefPart[1].toUpperCase()] || '?';
+
+    let partParams = fieldDefPart[2];
+    let closingBraceIndex = 0;
+
+    // FIXME: don't use a loop
+    // FIXME: separate maxbins from nestedPropertyParent and support other bin properties
+    for (let nestedPropertyParent of ['maxbins', 'scale', 'sort', 'axis', 'legend']) {
+      let nestedPropertyParentIndex = partParams.indexOf(nestedPropertyParent, closingBraceIndex);
+      if (nestedPropertyParentIndex !== -1) {
+        if (partParams[nestedPropertyParentIndex + nestedPropertyParent.length + 1] === '{') {
+          let openingBraceIndex = nestedPropertyParentIndex + nestedPropertyParent.length + 1;
+          closingBraceIndex = getClosingBraceIndex(openingBraceIndex, partParams);
+          encQ[nestedPropertyParent] = JSON.parse(partParams.substring(openingBraceIndex, closingBraceIndex + 1));
+
+        } else {
+          // Substring until the next comma (or end of the string)
+          let nextCommaIndex = partParams.indexOf(',', nestedPropertyParentIndex + nestedPropertyParent.length);
+          if (nextCommaIndex === -1) {
+            nextCommaIndex = partParams.length;
+          }
+
+          let parsedValue = JSON.parse(
+            partParams.substring(
+              nestedPropertyParentIndex + nestedPropertyParent.length + 1,
+              nextCommaIndex
+            )
+          );
+
+          if (nestedPropertyParent === 'maxbins') {
+            encQ.bin['maxbins'] = parsedValue;
+          } else {
+            encQ[nestedPropertyParent] = parsedValue;
+          }
+        }
+      }
+    }
+
+    return encQ;
+  }
+
+  // TODO(https://github.com/uwdata/compassql/issues/259):
+  // Extend this to support nested braces and brackets
+  export function getClosingBraceIndex(openingBraceIndex: number, str: string): number {
+    for (let i = openingBraceIndex; i < str.length; i++) {
+      if (str[i] === '}') {
+        return i;
+      }
+    }
+  }
+
+  export function fn(encQ: EncodingQuery, fieldDefShorthand: string): EncodingQuery {
+    // Aggregate, Bin, TimeUnit as enum spec case
+    if (fieldDefShorthand[0] === '?') {
+      let closingBraceIndex = getClosingBraceIndex(1, fieldDefShorthand);
+
+      let fnEnumIndex = JSON.parse(fieldDefShorthand.substring(1, closingBraceIndex + 1));
+
+      for (let encodingProperty in fnEnumIndex) {
+        if (isArray(fnEnumIndex[encodingProperty])) {
+          encQ[encodingProperty] = {enum: fnEnumIndex[encodingProperty]};
+        } else {
+          encQ[encodingProperty] = fnEnumIndex[encodingProperty];
+        }
+      }
+
+      return rawFieldDef(encQ,
+        splitWithTail(fieldDefShorthand.substring(closingBraceIndex + 2, fieldDefShorthand.length), ',', 2)
```

Comment:
why `fieldDefShorthand.length`, not `fieldDefShorthand.length-1`?  

Don't you have to exclude `)` at the end?   (If the answer is you handle it inside `rawFieldDef` then it's wrong because rawFieldDef shouldn't assume knowledge of an outsider method.  This would probably be a kind of leaky abstraction.